index.js

```diff
@@ -1053,22 +1057,51 @@ class Hypercore extends EventEmitter {
   }
 
   _updateEncryption () {
+    this.core._ensureEncryptionContext()
+
     const e = this.encryption
-    if (HypercoreEncryption.isHypercoreEncryption(e)) return
+    const ctx = this.core.encryptionContext
 
-    this.encryption = this._getLegacyEncryption(e.key, e.block)
+    if (!HypercoreEncryption.isHypercoreEncryption(e)) {
+      this.encryption = this._getEncryptionProvider(e.key, e.block)
+    }
+
+    if (!ctx.key) ctx.key = this.key
+    if (!ctx.manifest) ctx.manifest = this.manifest
+
+    if (this.padding === 0) {
+      try {
+        this.padding = this.encryption.paddingLength(ctx)
+      } catch {
+        // will retry later
+      }
+    }
 
     if (e === this.core.encryption) this.core.encryption = this.encryption
   }
 
-  _getLegacyEncryption (encryptionKey, block) {
+  _getEncryptionProvider (encryptionKey, block) {
     if (!encryptionKey) return null
 
     const blockKey = block
       ? encryptionKey
       : getLegacyBlockKey(this.key, encryptionKey, this.core.compat)
 
-    return HypercoreEncryption.createLegacyProvider(blockKey)
+    this.core.encryptionContext.blockKey = blockKey
+
+    return new HypercoreEncryption(crypto.hash(blockKey), {
+      getBlockKey: this._getBlockKey.bind(this) // only set once per session
+    })
```

Comment:
shouldnt we make a proper class for this so the block key can just be passed in the constructor and we dont need the runtime bind here?